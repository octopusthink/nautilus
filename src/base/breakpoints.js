// These are all the breakpoint sizes for use within your system.
// We currently use pixel values, but ems work too.

export const sizes = {
  xs: '414px',
  sm: '768px',
  md: '960px',
  lg: '1040px',
  xl: '1440px',
  xxl: '1640px',
};

// This variable returns an object containing all available media query wrappers for use within styled components
//
// Example Usage:
//
// ${mediaQueries.sm} {
//   ...css rules
// }
//

const mediaQueries = Object.keys(sizes).reduce((acc, bp) => ({
  ...acc,
  [bp]: `@media (min-width: ${sizes[bp]})`,
}), {});

export default mediaQueries;
