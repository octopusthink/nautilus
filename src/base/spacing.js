// ----------------------------------------
// Absolute Horizontal and Vertical Spacing
//
// https://ether.thescenery.co/spacing/
//
//
// Ether used the core typeface to generate sizes for the system spacing.
// fontCapHeight is defined as the x-height of the base0 font size
// ----------------------------------------

// Create spacing sizes tied to typography

const fontCapHeight = 0.9; // Cap-height of the base0 font size
const sizeModifier = fontCapHeight * 10; // used to convert to px in the end

// Calculate individual spacer heights/widths
const spacerNone = 0;
const spacer2xs = Math.round(sizeModifier * 0.5); // 5px
const spacerXs = sizeModifier; // 9px
const spacerSm = Math.round(sizeModifier * 1.5); // 14px
const spacerMd = Math.round(sizeModifier * 2.5); // 23px
const spacerLg = Math.round(sizeModifier * 4); // 36px
const spacerXl = Math.round(sizeModifier * 6.5); // 59px
const spacer2xl = Math.round(sizeModifier * 10.5); // 95px

// Spacer Options
const spacerSizes = {
  'none': `${spacerNone}px`,
  '2xsmall': `${spacer2xs}px`,
  'xsmall': `${spacerXs}px`,
  'small': `${spacerSm}px`,
  'medium': `${spacerMd}px`,
  'large': `${spacerLg}px`,
  'xlarge': `${spacerXl}px`,
  '2xlarge': `${spacer2xl}px`,
};

export default spacerSizes;
