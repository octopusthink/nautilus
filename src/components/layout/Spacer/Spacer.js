import PropTypes from 'prop-types';
import React from 'react';
import styled, { css } from 'styled-components';

import spacerSizes from '../../../base/spacing';
import mediaQueries from '../../../base/breakpoints';

export const Spacer = ({ debug, horizontal, responsiveSizes, size, ...others }) => {
  const spacerBackground = debug ? 'pink' : 'transparent';
  const spacerPadding = horizontal ? 'padding-right' : 'padding-bottom';
  const spacerDisplay = horizontal ? 'inline-block' : 'block';
  const spacerHeight = (horizontal && debug) ? '5px' : '0';

  // mutate the padding based on mediaQuery settings
  const paddingSize = spacerSizes[size] || '0px';
  let responsivePaddingSizes = css``;

  if (responsiveSizes) {
    responsivePaddingSizes = Object.entries(responsiveSizes).map((respSize) => {
      const [bp, sp] = respSize;
      return css`
        ${mediaQueries[bp]} {
          ${spacerPadding}: ${spacerSizes[sp]};
        }
      `;
    });
  }

  const StyledSpacer = styled.div`
    background-color: ${spacerBackground};
    display: ${spacerDisplay};
    height: ${spacerHeight};
    ${spacerPadding}: ${paddingSize};
    ${responsivePaddingSizes};
    vertical-align: middle;
  `;

  return (
    <StyledSpacer {...others} />
  );
};

export const SpacerSizes = Object.keys(spacerSizes);

Spacer.propTypes = {
  /** If set, adds a light background color to the spacer for debugging purposes */
  debug: PropTypes.bool,
  /** Sets to horizontal spacing mode */
  horizontal: PropTypes.bool,
  /** Object containing breakpoint-spacer size key-value pairs. See below for example. */
  responsiveSizes: PropTypes.shape({
    xs: PropTypes.oneOf(SpacerSizes),
    sm: PropTypes.oneOf(SpacerSizes),
    md: PropTypes.oneOf(SpacerSizes),
    lg: PropTypes.oneOf(SpacerSizes),
    xl: PropTypes.oneOf(SpacerSizes),
    xxl: PropTypes.oneOf(SpacerSizes),
  }),
  /** Size of the spacer. Options are 'none', '2xsmall', 'xsmall', 'small', 'medium', 'large', 'xlarge', and '2xlarge' */
  size: PropTypes.oneOf(SpacerSizes),
};

Spacer.defaultProps = {
  size: 'medium',
};
