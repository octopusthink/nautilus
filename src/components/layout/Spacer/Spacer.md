### Vertical Spacer
```js
<Spacer debug size="medium" />
```

### Horizontal Spacer
```js
const { Button } = require('../../input/Button/Button');
<span>
  <Button>Button 1</Button>
  <Spacer horizontal debug size="large" />
  <Button>Button 2</Button>
</span>
```

### Responsive Spacer
It's possible to add a `responsiveSizes` configuration object to change the size of the Spacer at different viewport sizes. Setting this value will override the `size` prop value.
```jsx static
const responsiveSizes = {
  xs: '2xsmall',
  sm: 'xlarge',
  lg: '2xlarge',
};

return (<Spacer responsiveSizes={responsiveSizes} />);
```
```js
<Spacer debug responsiveSizes={{ xs: '2xsmall', sm: 'xlarge', lg: '2xlarge' }} />
```
