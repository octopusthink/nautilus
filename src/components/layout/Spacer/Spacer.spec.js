import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';

import { Spacer } from './Spacer';
import spacerSizes from '../../../base/spacing';
import { sizes } from '../../../base/breakpoints';

describe('Spacer Component', () => {
  it('should match the snapshot', () => {
    const subject = shallow(<Spacer />);
    expect(subject).toMatchSnapshot();
  });

  it('should render inline-block if set to horizontal', () => {
    const subject = renderer.create(<Spacer horizontal />).toJSON();
    expect(subject).toMatchSnapshot();
    expect(subject).toHaveStyleRule('display', 'inline-block');
  });

  it('should receive correct padding based on size selection', () => {
    const subject = renderer.create(<Spacer size="large" />).toJSON();
    const mediumSpacerSize = spacerSizes.large;
    expect(subject).toMatchSnapshot();
    expect(subject).toHaveStyleRule('padding-bottom', mediumSpacerSize);
  });

  context('when provided a responsiveSizes object', () => {
    it('contains the correct media queries', () => {
      const responsiveSizes = {
        xs: '2xsmall',
        lg: '2xlarge',
      };

      const subject = renderer.create(<Spacer responsiveSizes={responsiveSizes} />).toJSON();
      expect(subject).toMatchSnapshot();
      expect(subject).toHaveStyleRule('padding-bottom', spacerSizes['2xsmall'], {
        media: `(min-width: ${sizes.xs})`,
      });

      expect(subject).toHaveStyleRule('padding-bottom', spacerSizes['2xlarge'], {
        media: `(min-width: ${sizes.lg})`,
      });
    });
  });
});
