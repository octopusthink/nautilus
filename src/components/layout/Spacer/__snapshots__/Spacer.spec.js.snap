// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spacer Component should match the snapshot 1`] = `<ForwardRef />`;

exports[`Spacer Component should receive correct padding based on size selection 1`] = `
.c0 {
  background-color: transparent;
  display: block;
  height: 0;
  padding-bottom: 36px;
  vertical-align: middle;
}

<div
  className="c0"
/>
`;

exports[`Spacer Component should render inline-block if set to horizontal 1`] = `
.c0 {
  background-color: transparent;
  display: inline-block;
  height: 0;
  padding-right: 23px;
  vertical-align: middle;
}

<div
  className="c0"
/>
`;

exports[`Spacer Component when provided a responsiveSizes object contains the correct media queries 1`] = `
.c0 {
  background-color: transparent;
  display: block;
  height: 0;
  padding-bottom: 23px;
  vertical-align: middle;
}

@media (min-width:414px) {
  .c0 {
    padding-bottom: 5px;
  }
}

@media (min-width:1040px) {
  .c0 {
    padding-bottom: 95px;
  }
}

<div
  className="c0"
/>
`;
